src/main/java/com/uid2/shared/auth/AuthorizableStore.java

```diff
@@ -1,77 +1,154 @@
+    private final LoadingCache<String, String> keyToHashCache;
 
     public AuthorizableStore() {
         this.authorizables = new AtomicReference<>(new AuthorizableStoreSnapshot(new ArrayList<>()));
+        this.keyToHashCache = createCache();
     }
 
     public void refresh(Collection<T> authorizablesToRefresh) {
         authorizables.set(new AuthorizableStoreSnapshot(authorizablesToRefresh));
     }
 
-    public T getFromKey(String key) {
+    public T getAuthorizableByKey(String key) {
         AuthorizableStoreSnapshot latest = authorizables.get();
-        // TODO: add caching
-        for (byte[] salt : latest.getSalts()) {
+
+        T authorizable;
+
+        authorizable = getAuthorizableBySiteId(latest, key);
+        if (authorizable != null) {
+            return authorizable;
+        }
+
+        String cachedHash = keyToHashCache.get(key);
+        if (cachedHash != null) {
+            return cachedHash.isBlank() ? null : latest.getAuthorizableByHash(wrapHashToByteBuffer(cachedHash));
+        }
+
+        authorizable = hashAndCompareKeyAgainstAuthorizables(latest, key);
+        keyToHashCache.put(key, authorizable == null ? "" : authorizable.getKeyHash());
+
+        return authorizable;
+    }
+
+    public T getAuthorizableByHash(String hash) {
+        ByteBuffer hashBytes = wrapHashToByteBuffer(hash);
+        if (hashBytes == null) {
+            return null;
+        }
+
+        AuthorizableStoreSnapshot latest = authorizables.get();
+        return latest.getAuthorizableByHash(hashBytes);
+    }
+
+    private T getAuthorizableBySiteId(AuthorizableStoreSnapshot snapshot, String key) {
+        Integer siteId = getSiteIdFromKey(key);
+
+        if (siteId != null) {
+            T authorizable = snapshot.getAuthorizableBySiteId(siteId);
+            if (authorizable == null) {
+                return null;
+            }
+
+            byte[] hash = KEY_HASHER.hashKey(key, convertBase64StringToBytes(authorizable.getKeySalt()));
+            return snapshot.getAuthorizableByHash(ByteBuffer.wrap(hash));
+        }
+
+        return null;
+    }
+
+    private T hashAndCompareKeyAgainstAuthorizables(AuthorizableStoreSnapshot snapshot, String key) {
+        for (byte[] salt : snapshot.getSalts()) {
             byte[] keyHash = KEY_HASHER.hashKey(key, salt);
-            T authorizable = latest.getAuthorizables().get(ByteBuffer.wrap(keyHash));
+            T authorizable = snapshot.getAuthorizableByHash(ByteBuffer.wrap(keyHash));
             if (authorizable != null) {
                 return authorizable;
             }
         }
         return null;
     }
 
-    public T getFromKeyHash(String keyHash) {
-        AuthorizableStoreSnapshot latest = authorizables.get();
+    private static LoadingCache<String, String> createCache() {
+        return Caffeine.newBuilder()
+                .maximumSize(CACHE_MAX_SIZE)
+                .recordStats()
```

Comment:
No way to get to those stats for now - until we export them, let's not calculate them.